test/common/stats/thread_local_store_test.cc

```diff
@@ -450,6 +450,38 @@ TEST_F(StatsThreadLocalStoreTest, OverlappingScopes) {
   tls_.shutdownThread();
 }
 
+// Demonstrates that counters, gauges, and indicators are all mixed together in
+// the shared memory, and not separated by type; only the name matters.
+TEST_F(StatsThreadLocalStoreTest, SameNameDifferentType) {
+  InSequence s;
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
+  EXPECT_CALL(*alloc_, alloc(_)).Times(4);
+
+  Counter& c1 = store_->counter("samename");
+  EXPECT_EQ(&c1, &store_->counter("samename"));
+  Gauge& g1 = store_->gauge("samename");
+  EXPECT_EQ(&g1, &store_->gauge("samename"));
+  c1.add(5);
+  EXPECT_EQ(5UL, c1.value());
+  g1.add(3);
+  EXPECT_EQ(8UL, c1.value());
+
+  Gauge& g2 = store_->gauge("samename2");
```

Comment:
Is this the test you added per my suggestion? This does cover the question I wanted to answer, so thanks. However, it also locks in behavior that I think might not really be what we want once you remove hot-restart, where the underlying storage is shared independent of the type.

I think that's OK for now as you did find a way to test the functionality, but maybe we should comment that this is really an artifact of the current implementation and we might prefer to disallow Counter, Gauge, and BoolIndicator with the same name to share backing store. I think in the future we might want to make this illegal.

Of course I'm afraid of http://www.hyrumslaw.com/ here...

@mattklein123 can you share whether having a counter and gauge named the same thing and sharing backing store is intended behavior?

If we don't want to guarantee this behavior let's leave a note here that we are taking advantage of it temporarily only until the hot-restart-via-rpc PR is in.

Reply:
> @mattklein123 can you share whether having a counter and gauge named the same thing and sharing backing store is intended behavior?

I think I would call this an artifact of the current implementation and it's probably fine to change. So I'm fine just putting a note in somewhere to this effect.

Reply:
Nice, note added. I am glad of this suggestion - I was surprised to learn it worked this way!